python/lsst/ts/m2gui/controltab/tab_diagnostics.py

```diff
@@ -33,7 +44,689 @@ class TabDiagnostics(TabDefault):
+
+        # Utility data
+        self._power_calibrated = self._create_labels_power()
+        self._power_raw = self._create_labels_power()
+
+        self._force_error_tangent = {
+            "a1": create_label(),
+            "a2": create_label(),
+            "a3": create_label(),
+            "a4": create_label(),
+            "a5": create_label(),
+            "a6": create_label(),
+            "total_weight": create_label(),
+            "tangent_sum": create_label(),
+        }
+
+        self._button_reboot = set_button(
+            "Reboot Controller", self._callback_reboot_controller
+        )
+
+        self._digital_status_input = self._create_indicators_digital_status(
+            self._get_list_digital_status_input()
+        )
+        self._digital_status_output = self._create_indicators_digital_status(
+            self._get_list_digital_status_output()
+        )
+
+        self._digital_status_control = self._create_controls_digital_status(
+            len(self._get_list_digital_status_output())
+        )
+
+        self._set_widget_and_layout()
+
+        self._set_signal_utility(self.model.utility_monitor.signal_utility)
+        self._set_signal_detailed_force(
+            self.model.utility_monitor.signal_detailed_force
+        )
+
+    def _create_labels_power(self):
+        """Create the labels of power.
+
+        Returns
+        -------
+        `dict [PySide2.QtWidgets.QLabel]`
+            Labels of the power. The key is the specific power's name.
+        """
+
+        return {
+            "power_voltage_motor": create_label(),
+            "power_current_motor": create_label(),
+            "power_voltage_communication": create_label(),
+            "power_current_communication": create_label(),
+        }
+
+    @Slot()
+    def _callback_reboot_controller(self):
+        """Callback of the reboot-cell-controller button. This will ask the
+        user to confirm this command again."""
+
+        dialog = QMessageBox()
+
+        dialog.setText("This will reboot the cell controller.")
+        dialog.setInformativeText("Are you sure to reboot the controller?")
+
+        dialog.setStandardButtons(QMessageBox.Ok | QMessageBox.Cancel)
+        dialog.setDefaultButton(QMessageBox.Cancel)
+
+        decision = dialog.exec()
+        if decision == QMessageBox.Ok:
+            run_command(self.model.reboot_controller)
+
+    def _create_indicators_digital_status(self, list_digital_status):
+        """Create the indicators of digital status.
+
+        Parameters
+        ----------
+        list_digital_status : `list`
+            List of the digital status.
+
+        Returns
+        -------
+        indicators_status : `list [PySide2.QtWidgets.QPushButton]`
+            Indicators of the digital status.
+        """
+
+        indicators_status = list()
+        for status in list_digital_status:
+            indicator = set_button(status, None, is_indicator=True, is_adjust_size=True)
```

Comment:
If you choose to inherit instead of composing, you can just create an extra class that will _update_indicator_color as part of its logic (in __init__, and then you can use the method as needed). It makes your code better, as you don't have to care about details, and still see what the indicator is doing. With insisting on composition, you are creating a particular button, which is an indicator, but because you don't inherit and don't use own classes, you have to add a method to the class which doesn't have any logic common with indicator functionality to change its color. You cannot reuse that code, have to write again the same method in another class (or move that to famous utils, instead of localizing code in a custom class, which will do the behavior). You have been warned, and I hope you now see where composing leads - it creates code, particularly for GUIs, which isn't modular or is modular at a very high cost. There is a reason why Qt uses inhertance everywhere it can, and almost never compose.

Reply:
I think the problem here is that I do not know yet how many different colors I will use for `_update_indicator_color()` in the final. Right now, for the different table, I have different color sets for the indicators ("red/green", "green/gray", etc.). The reason is that I tried to duplicate the existed LabVIEW EUI as much as possible. After I finished the EUI, as what you said, I should consider this part and do the related refactoring. Thanks for the suggestion!